core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -256,74 +384,73 @@ class ProducerIdMapping(val config: LogConfig,
     }
   }
 
-  private def isExpired(currentTimeMs: Long, producerIdEntry: ProducerIdEntry) : Boolean =
-    currentTimeMs - producerIdEntry.timestamp >= maxPidExpirationMs
-
+  private def isExpired(currentTimeMs: Long, producerIdEntry: ProducerIdEntry): Boolean =
+    producerIdEntry.currentTxnFirstOffset.isEmpty && currentTimeMs - producerIdEntry.timestamp >= maxPidExpirationMs
 
-  def removeExpiredPids(currentTimeMs: Long) {
-    pidMap.retain { case (pid, lastEntry) =>
+  /**
+   * Expire any PIDs which have been idle longer than the configured maximum expiration timeout.
+   */
+  def removeExpiredProducers(currentTimeMs: Long) {
+    producers.retain { case (pid, lastEntry) =>
       !isExpired(currentTimeMs, lastEntry)
     }
   }
 
   /**
    * Truncate the PID mapping to the given offset range and reload the entries from the most recent
-   * snapshot in range (if there is one).
+   * snapshot in range (if there is one). Note that the log end offset is assumed to be less than
+   * or equal to the high watermark.
```

Comment:
Hmm, not sure about the comment on HW. With KIP-101, typically the truncation point is >= the local HW.